x/superfluid/keeper/superfluid_test.go

```diff
@@ -0,0 +1,182 @@
+package keeper_test
+
+import (
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+	appparams "github.com/osmosis-labs/osmosis/v7/app/params"
+	"github.com/osmosis-labs/osmosis/v7/x/epochs"
+	epochstypes "github.com/osmosis-labs/osmosis/v7/x/epochs/types"
+	lockuptypes "github.com/osmosis-labs/osmosis/v7/x/lockup/types"
+	"github.com/osmosis-labs/osmosis/v7/x/superfluid/keeper"
+	"github.com/osmosis-labs/osmosis/v7/x/superfluid/types"
+)
+
+// TODO: add more test cases
+func (suite *KeeperTestSuite) TestSuperfluidFlow() {
+	testCases := []struct {
+		name               string
+		validatorStats     []stakingtypes.BondStatus
+		superDelegation    superfluidDelegation
+		expInterDelegation sdk.Dec
+	}{
+		{
+			"with single validator and single superfluid delegation",
+			[]stakingtypes.BondStatus{stakingtypes.Bonded},
+			superfluidDelegation{0, "gamm/pool/1"},
+			sdk.NewDec(19000000), // 95% x 2 x 1000000
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		suite.Run(tc.name, func() {
+			suite.SetupTest()
+
+			// we set suite.ctx so that other suite methods are ran in the correct block height and block time
+			now := time.Now()
+			suite.ctx = suite.ctx.WithBlockHeight(1).WithBlockTime(now)
+			// run initial setup for epoch, starting with deleting old epoch information
+			epochInfos := suite.app.EpochsKeeper.AllEpochInfos(suite.ctx)
+			for _, epochInfo := range epochInfos {
+				suite.app.EpochsKeeper.DeleteEpochInfo(suite.ctx, epochInfo.Identifier)
+			}
+
+			epochInfo := epochstypes.EpochInfo{
+				Identifier:            "day",
+				StartTime:             suite.ctx.BlockTime(),
+				Duration:              time.Hour * 24,
+				CurrentEpoch:          0,
+				CurrentEpochStartTime: time.Time{},
+				EpochCountingStarted:  false,
+			}
+			suite.app.EpochsKeeper.SetEpochInfo(suite.ctx, epochInfo)
+
+			epochs.BeginBlocker(suite.ctx, *suite.app.EpochsKeeper)
```

Comment:
Would be ideal if we can switch to App.BeginBlock() / app.EndBlock(), so we can understand whats the pipeline of all the moving components

Reply:
(Somewhat worried that the order of calls in here will be different than what we actually use)

Reply:
good catch!